brainiak/utils/fmrisim.py

```diff
@@ -678,90 +681,106 @@ def apply_signal(signal_function,
-                    countSqDer[2] += 1
-
-        # What is the FWHM for each dimension
-        fwhm3 = np.sqrt(np.divide(4 * np.log(2), np.divide(sum_sq_der,
-                                                           countSqDer)))
-        fwhm[tr_counter] = np.prod(fwhm3) ** (1 / 3)  # Calculate the average
+    # Preset
+    v_count = 0
+    v_sum = 0
+    v_sq = 0
+
+    d_sum = [0.0, 0.0, 0.0]
+    d_sq = [0.0, 0.0, 0.0]
+    d_count = [0, 0, 0]
+
+    # Pull out all the voxel coordinates
+    coordinates = list(product(range(dimensions[0]),
+                               range(dimensions[1]),
+                               range(dimensions[2])))
+
+    # Find the sum of squared error for the non-masked voxels in the brain
+    for i in list(range(len(coordinates))):
+
+        # Pull out this coordinate
+        x, y, z = coordinates[i]
+
+        # Is this within the mask?
+        if mask[x, y, z] > 0:
+
+            # Find the the volume sum and squared values
+            v_count += 1
+            v_sum += np.abs(volume[x, y, z])
+            v_sq += volume[x, y, z] ** 2
+
+    # Get the volume variance
+    v_var = (v_sq - ((v_sum ** 2) / v_count)) / (v_count - 1)
+
+    for i in list(range(len(coordinates))):
+
+        # Pull out this coordinate
+        x, y, z = coordinates[i]
+
+        # Is this within the mask?
+        if mask[x, y, z] > 0:
+            # For each xyz dimension calculate the squared
+            # difference of this voxel and the next
+
+            inrange = (x < dimensions[0] - 1)
+            inmask = inrange and (mask[x + 1, y, z] > 0)
+            included = inmask and (~np.isnan(volume[x + 1, y, z]))
+            if included:
+                d_sum[0] += volume[x, y, z] - volume[x + 1, y, z]
+                d_sq[0] += (volume[x, y, z] - volume[x + 1, y, z]) ** 2
+                d_count[0] += 1
+
+            inrange = (y < dimensions[1] - 1)
+            inmask = inrange and (mask[x, y + 1, z] > 0)
+            included = inmask and (~np.isnan(volume[x, y + 1, z]))
+            if included:
+                d_sum[1] += volume[x, y, z] - volume[x, y + 1, z]
+                d_sq[1] += (volume[x, y, z] - volume[x, y + 1, z]) ** 2
+                d_count[1] += 1
+
+            inrange = (z < dimensions[2] - 1)
+            inmask = inrange and (mask[x, y, z + 1] > 0)
+            included = inmask and (~np.isnan(volume[x, y, z + 1]))
+            if included:
+                d_sum[2] += volume[x, y, z] - volume[x, y, z + 1]
+                d_sq[2] += (volume[x, y, z] - volume[x, y, z + 1]) ** 2
+                d_count[2] += 1
+
+    # Find the variance
+    d_var = np.divide((d_sq - np.divide(np.power(d_sum, 2),
+                                        d_count)), (np.add(d_count, -1)))
+
+    o_var = np.divide(-1, (4 * np.log(1 - (0.5 * d_var / v_var))))
+    fwhm3 = np.sqrt(o_var) * 2 * np.sqrt(2 * np.log(2))
+    fwhm = np.prod(np.multiply(fwhm3, voxel_size)) ** (1 / 3)
 
     return fwhm
```

Comment:
To run this on a single TR with 64x64x48 voxels takes 1.002s. In calc_noise I subsample the trs to avoid this taking more than 100 TRs. At the moment this is a sufficient work around.